### PR TITLE
kraken: reuse protobuf response

### PR DESCRIPTION
**Pull request by kadhikari**
concerned ticket: http://jira.canaltp.fr/browse/ITI-248
Goal: save allocation time

### CODE REVIEW

File: source/kraken/tests/direct_path_test.cpp [hunk 1]
```diff
@@ -74,30 +75,44 @@ BOOST_AUTO_TEST_CASE(direct_path_test) {
 
     // walking
     sn_params->set_origin_mode("walking");
-    auto res = w.direct_path(req);
+    //auto * data_ptr = data_manager.get_data().get();
```
Comment (xlqian) [inline source/kraken/tests/direct_path_test.cpp]: comments are still useful?

File: source/ptreferential/ptreferential_api.cpp [hunk 2]
```diff
@@ -41,124 +41,134 @@ namespace pt = boost::posix_time;
...
                              const int startPage,
                              const int count,
                              const boost::optional<boost::posix_time::ptime>& since,
                              const boost::optional<boost::posix_time::ptime>& until,
-                             const type::Data& data,
-                             const boost::posix_time::ptime& current_datetime) {
+                             const type::Data& data) {
     type::Indexes final_indexes;
-    pbnavitia::Response pb_response;
+    //pbnavitia::Response pb_response;
```
Comment (xlqian) [inline source/ptreferential/ptreferential_api.cpp]: to remove?

File: source/kraken/kraken_zmq.h [hunk 3]
```diff
@@ -94,13 +89,12 @@ inline void doWork(zmq::context_t& context,
         }
 
         pbnavitia::Request pb_req;
-        pbnavitia::Response result;
         pt::ptime start = pt::microsec_clock::universal_time();
         pbnavitia::API api = pbnavitia::UNKNOWN_API;
-        if(!pb_req.ParseFromArray(request.data(), request.size())){
+        if(!pb_req.ParseFromArray(request.data(), request.size())){
```
Comment (TeXitoi) [inline source/kraken/kraken_zmq.h]: training spaces

File: source/routing/raptor_api.cpp [hunk 4]
```diff
@@ -1327,27 +1318,26 @@ static DateTime make_isochrone_date (const DateTime& init_dt,
...
-                                            const std::vector<std::string>& forbidden,
-                                            bool clockwise,
-                                            const nt::RTLevel rt_level,
-                                            georef::StreetNetwork & worker,
-                                            const double& speed) {
-
-    PbCreator pb_creator(raptor.data, current_datetime, null_time_period);
+void make_graphical_isochrone(navitia::PbCreator& pb_creator,
+                              RAPTOR &raptor,
+                              type::EntryPoint center,
```
Comment (xlqian) [inline source/routing/raptor_api.cpp]: can it be a const ref? or ref?

File: source/routing/raptor_api.cpp [hunk 5]
```diff
@@ -1358,39 +1348,33 @@ pbnavitia::Response make_graphical_isochrone(RAPTOR &raptor,
...
-                                  const nt::RTLevel rt_level,
-                                  georef::StreetNetwork & worker,
-                                  const double& speed,
-                                  const navitia::type::Mode_e mode,
-                                  const uint32_t resolution) {
-
-    PbCreator pb_creator(raptor.data, current_datetime, null_time_period);
+void make_heat_map(navitia::PbCreator& pb_creator,
+                   RAPTOR &raptor,
+                   type::EntryPoint center,
```
Comment (xlqian) [inline source/routing/raptor_api.cpp]: const ref? ref?

File: source/time_tables/route_schedules.cpp [hunk 6]
```diff
@@ -353,31 +353,31 @@ void route_schedule(PbCreator& pb_creator, const std::string& filter,
...
             stop_points.push_back(vector_idx());
             for (const auto& jpp_idx : jp.jpps) {
-                const auto& jpp = pb_creator.data.dataRaptor->jp_container.get(jpp_idx);
+                const auto& jpp = pb_creator.data->dataRaptor->jp_container.get(jpp_idx);
                 stop_points.back().push_back(jpp.sp_idx.val);
             }
         }
         thermometer.generate_thermometer(stop_points);
-        auto  matrice = make_matrice(stop_times, thermometer, pb_creator.data);
+        auto  matrice = make_matrice(stop_times, thermometer, *pb_creator.data);
```
Comment (xlqian) [inline source/time_tables/route_schedules.cpp]: make_matrix :)  i know it's not your codes

File: source/kraken/tests/direct_path_test.cpp [hunk 7]
```diff
@@ -74,30 +75,44 @@ BOOST_AUTO_TEST_CASE(direct_path_test) {
...
+    //navitia::PbCreator pb_creator(data_ptr, boost::gregorian::not_a_date_time, null_time_period);
+    w.init_worker_data(data_manager.get_data(), boost::gregorian::not_a_date_time, null_time_period);
+    //w.pb_creator.init(data_ptr, boost::gregorian::not_a_date_time, null_time_period);
```
Comment (TeXitoi) [inline source/kraken/tests/direct_path_test.cpp]: comment cleanup needed.

File: source/kraken/tests/direct_path_test.cpp [hunk 8]
```diff
@@ -74,30 +75,44 @@ BOOST_AUTO_TEST_CASE(direct_path_test) {
...
+    auto res = w.pb_creator.get_response();
     BOOST_REQUIRE_EQUAL(res.journeys_size(), 1);
     BOOST_CHECK_EQUAL(res.journeys(0).sections_size(), 1);
     BOOST_CHECK_EQUAL(res.journeys(0).sections(0).street_network().path_items_size(), 3);
     BOOST_CHECK_EQUAL(res.journeys(0).durations().total(), 275);
 
     // bss
     sn_params->set_origin_mode("bss");
-    res = w.direct_path(req);
+    //w.pb_creator.init(data_ptr, boost::gregorian::not_a_date_time, null_time_period);
```
Comment (TeXitoi) [inline source/kraken/tests/direct_path_test.cpp]: comment cleanup

File: source/kraken/worker.cpp [hunk 9]
```diff
@@ -186,9 +186,25 @@ static std::string get_string_status(const boost::shared_ptr<const nt::Data>& da
...
+        case pbnavitia::DEPARTURE_BOARDS: result = request.next_stop_times().disable_geojson(); break;
+        case pbnavitia::PTREFERENTIAL: result = request.ptref().disable_geojson(); break;
+    default: result = false; break;
+    }
+    return result;
+}
+
+void Worker::geo_status() {
+    auto status = this->pb_creator.mutable_geo_status();
+    //auto status = result.mutable_geo_status();
```
Comment (TeXitoi) [inline source/kraken/worker.cpp]: comment cleanup

**Review by xlqian — COMMENTED** [review]
good job!

File: source/kraken/worker.cpp [hunk 10]
```diff
@@ -665,176 +663,154 @@ navitia::JourneysArg Worker::fill_journeys(const pbnavitia::JourneysRequest &req
...
+                                                  bt::from_time_t(request.since_datetime())),
+                             boost::make_optional(request.has_until_datetime(),
+                                                  bt::from_time_t(request.until_datetime())),
+                             *data);
 }
 
 
-boost::optional<pbnavitia::Response> Worker::set_journeys_args(const pbnavitia::JourneysRequest& request,
-                                                               const boost::posix_time::ptime& current_datetime,
+bool Worker::set_journeys_args(const pbnavitia::JourneysRequest& request,
```
Comment (TeXitoi) [inline source/kraken/worker.cpp:749]: A small comment explaining the return value.

File: source/ptreferential/ptreferential_api.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/ptreferential/ptreferential_api.cpp]: comment

File: source/time_tables/tests/departure_boards_test.cpp [hunk 11]
```diff
@@ -277,12 +287,13 @@ BOOST_AUTO_TEST_CASE(departureboard_test_with_impacts) {
     BOOST_REQUIRE_EQUAL(impact->messages(0).text(), "Disruption on stop_point stop1");
 
     // stop_schedules on stop1 with rtLevel = RealTime
-    navitia::PbCreator pb_creator1(*(b.data), "20150616T080000"_dt, null_time_period);
-    pb_creator.now = "20150626T110000"_dt;
-    departure_board(pb_creator1, "stop_point.uri=stop1", {}, {}, d("20150615T094500"), 43200, 0,
+    pb_creator.init(data_ptr, "20150616T080000"_dt, null_time_period);
+    //navitia::PbCreator pb_creator1(data_ptr, "20150616T080000"_dt, null_time_period);
```
Comment (TeXitoi) [inline source/time_tables/tests/departure_boards_test.cpp]: comment

File: source/type/pb_converter.cpp [hunk 12]
```diff
@@ -1761,10 +1761,13 @@ void PbCreator::fill_pb_error(const pbnavitia::Error::error_id id, const std::st
...
 }
 
-pbnavitia::Response PbCreator::get_response(){
+const pbnavitia::Response& PbCreator::get_response(){
     Filler(0, DumpMessage::No, *this).fill_pb_object(contributors, response.mutable_feed_publishers());
+    contributors.clear();
     Filler(0, DumpMessage::No, *this).fill_pb_object(impacts, response.mutable_impacts());
-    return std::move(response);
+    impacts.clear();
+    //return std::move(response);
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: comment

**Review by TeXitoi — CHANGES_REQUESTED** [review]
:+1:

File: source/kraken/tests/direct_path_test.cpp (hunk 1 above)
Comment (kadhikari) [reply source/kraken/tests/direct_path_test.cpp]: ok

File: source/ptreferential/ptreferential_api.cpp (hunk 2 above)
Comment (kadhikari) [reply source/ptreferential/ptreferential_api.cpp]: ok

File: source/time_tables/route_schedules.cpp (hunk 6 above)
Comment (kadhikari) [reply source/time_tables/route_schedules.cpp]: done

File: source/kraken/kraken_zmq.h (hunk 3 above)
Comment (kadhikari) [reply source/kraken/kraken_zmq.h]: ok

File: source/kraken/tests/direct_path_test.cpp (hunk 7 above)
Comment (kadhikari) [reply source/kraken/tests/direct_path_test.cpp]: done

File: source/kraken/tests/direct_path_test.cpp (hunk 8 above)
Comment (kadhikari) [reply source/kraken/tests/direct_path_test.cpp]: done

File: source/kraken/worker.cpp (hunk 9 above)
Comment (kadhikari) [reply source/kraken/worker.cpp]: done

File: source/ptreferential/ptreferential_api.cpp (hunk 2 above)
Comment (kadhikari) [reply source/ptreferential/ptreferential_api.cpp]: done

File: source/time_tables/tests/departure_boards_test.cpp (hunk 11 above)
Comment (kadhikari) [reply source/time_tables/tests/departure_boards_test.cpp]: done

File: source/type/pb_converter.cpp (hunk 12 above)
Comment (kadhikari) [reply source/type/pb_converter.cpp]: done

**Review by TeXitoi — APPROVED** [review]

File: source/autocomplete/autocomplete_api.cpp [hunk 13]
```diff
@@ -183,22 +183,17 @@ static void update_quality(std::vector<Autocomplete<nt::idx_t>::fl_quality>& ac_
 }
 
 
-pbnavitia::Response autocomplete(const std::string &q,
+void autocomplete(navitia::PbCreator& pb_creator, const std::string &q,
```
Comment (pbougue) [inline source/autocomplete/autocomplete_api.cpp:186]: indent

File: source/autocomplete/autocomplete_api.h [hunk 14]
```diff
@@ -44,14 +45,12 @@ namespace type {
 namespace autocomplete {
 
 /** Trouve tous les objets définis par filter dont le nom contient q */
-pbnavitia::Response autocomplete(const std::string &q,
+void autocomplete(navitia::PbCreator& pb_creator, const std::string &q,
```
Comment (pbougue) [inline source/autocomplete/autocomplete_api.h:48]: indent

File: source/calendar/calendar_api.cpp [hunk 15]
```diff
@@ -35,60 +35,58 @@ www.navitia.io
...
-                              const boost::posix_time::ptime& current_datetime,
-                              const std::string &start_date,
-                              const std::string &end_date,
-                              const size_t depth,
-                              size_t count,
-                              size_t start_page,
-                              const std::string &filter,
-                              const std::vector<std::string>& forbidden_uris) {
+void calendars(navitia::PbCreator& pb_creator,
+          const navitia::type::Data &d,
```
Comment (pbougue) [inline source/calendar/calendar_api.cpp:39]: indent

File: source/calendar/calendar_api.cpp [hunk 16]
```diff
@@ -35,60 +35,58 @@ www.navitia.io
 
 namespace navitia { namespace calendar {
 
-pbnavitia::Response calendars(const navitia::type::Data &d,
-                              const boost::posix_time::ptime& current_datetime,
```
Comment (pbougue) [inline source/calendar/calendar_api.cpp:39]: are you sure it's not gonna prevent from debug (as you use `now` after), or imply some bugs to remove current_dt? (same question everywhere, so I guess you thinked about it)

File: source/calendar/calendar_api.h [hunk 17]
```diff
@@ -32,15 +32,16 @@ www.navitia.io
...
-pbnavitia::Response calendars(const navitia::type::Data &d,
-                              const boost::posix_time::ptime& current_datetime,
-                              const std::string &start_date,
-                              const std::string &end_date,
-                              const size_t depth,
-                              size_t count,
-                              size_t start_page, const std::string &filter,
-                              const std::vector<std::string>& forbidden_uris);
+void calendars(navitia::PbCreator& pb_creator,
+          const navitia::type::Data &d,
```
Comment (pbougue) [inline source/calendar/calendar_api.h:40]: indent

File: source/jormungandr/tests/ptref_tests.py [hunk 18]
```diff
@@ -92,13 +92,13 @@ def test_vj_default_depth(self):
         feed_publishers = response["feed_publishers"]
         for feed_publisher in feed_publishers:
             is_valid_feed_publisher(feed_publisher)
-        feed_publisher = feed_publishers[0]
+        feed_publisher = feed_publishers[1]
```
Comment (pbougue) [inline source/jormungandr/tests/ptref_tests.py:95]: Order is not really stable? Maybe not the matter of this PR but might be worth opening an issue?

File: source/kraken/tests/realtime_test.cpp [hunk 19]
```diff
@@ -1475,9 +1476,12 @@ BOOST_AUTO_TEST_CASE(traffic_reports_vehicle_journeys) {
     navitia::handle_realtime("trip_update_vj2", timestamp, trip_update_vj2, *b.data);
     navitia::handle_realtime("trip_update_vj3", timestamp, trip_update_vj3, *b.data);
 
-    const auto resp = navitia::disruption::traffic_reports(*b.data,
-                                                           boost::posix_time::from_iso_string("20150928T0830"),
-                                                           1, 10, 0, "", {});
+    auto * data_ptr = b.data.get();
+    navitia::PbCreator pb_creator(data_ptr, boost::posix_time::from_iso_string("20150928T0830"), null_time_period);
```
Comment (pbougue) [inline source/kraken/tests/realtime_test.cpp:1480]: is the dt really needed here? seems odd if it's the case.

File: source/kraken/worker.cpp [hunk 20]
```diff
@@ -291,41 +303,47 @@ void Worker::feed_publisher(pbnavitia::Response& response){
     }
 }
 
-void Worker::init_worker_data(const boost::shared_ptr<const navitia::type::Data> data){
+void Worker::init_worker_data(const boost::shared_ptr<const navitia::type::Data> data,
+                              const pt::ptime now,
```
Comment (pbougue) [inline source/kraken/worker.cpp:307]: I'd name it `current_datetime`

File: source/kraken/worker.cpp [hunk 21]
```diff
@@ -291,41 +303,47 @@ void Worker::feed_publisher(pbnavitia::Response& response){
...
-pbnavitia::Response Worker::traffic_reports(const pbnavitia::TrafficReportsRequest &request,
+void Worker::traffic_reports(const pbnavitia::TrafficReportsRequest &request,
                                             const boost::posix_time::ptime& current_datetime){
     const auto data = data_manager.get_data();
     std::vector<std::string> forbidden_uris;
     for(int i = 0; i < request.forbidden_uris_size(); ++i)
         forbidden_uris.push_back(request.forbidden_uris(i));
-    return navitia::disruption::traffic_reports(*data,
+    navitia::disruption::traffic_reports(this->pb_creator,
+                                                *data,
```
Comment (pbougue) [inline source/kraken/worker.cpp:346]: indent

File: source/kraken/worker.cpp [hunk 22]
```diff
@@ -424,33 +437,28 @@ pbnavitia::Response Worker::next_stop_times(const pbnavitia::NextStopTimeRequest
...
+                               coord,
                                request.distance(),
                                vector_of_pb_types(request),
                                request.filter(),
                                request.depth(),
                                request.count(),
                                request.start_page(),
-                               *data,
-                               current_datetime,
-                               disable_feedpublisher);
```
Comment (pbougue) [inline source/kraken/worker.cpp:453]: `disable_feedpublisher` not susceptible to be used downward?

File: source/kraken/worker.cpp [hunk 23]
```diff
@@ -541,20 +544,18 @@ pbnavitia::Response Worker::place_uri(const pbnavitia::PlaceUriRequest &request,
                 auto it_admin = data->geo_ref->admin_map.find(request.uri());
                 if(it_admin != data->geo_ref->admin_map.end()) {
                     pb_creator.fill(data->geo_ref->admins[it_admin->second], pb_creator.add_places(), request.depth());
-
                 }else{
-                    pb_creator.fill_pb_error(pbnavitia::Error::unable_to_parse,
+                    this->pb_creator.fill_pb_error(pbnavitia::Error::unable_to_parse,
                                              "Unable to parse : " + request.uri());
```
Comment (pbougue) [inline source/kraken/worker.cpp:549]: indent

File: source/kraken/worker.cpp [hunk 24]
```diff
@@ -670,176 +666,154 @@ navitia::JourneysArg Worker::fill_journeys(const pbnavitia::JourneysRequest &req
...
+    if (has_error) { return; }
+
     type::EntryPoint ep = arg.origins.empty() ? arg.destinations[0] : arg.origins[0];
     std::vector<DateTime> boundary_duration;
     for(int i = 0; i < request.boundary_duration_size(); ++i) {
         boundary_duration.push_back(request.boundary_duration(i));
     }
-    return navitia::routing::make_graphical_isochrone(*planner, current_datetime, ep, request_journey.datetimes(0),
+    navitia::routing::make_graphical_isochrone(this->pb_creator, *planner, ep, request_journey.datetimes(0),
                                                       boundary_duration, request_journey.max_transfers(),
```
Comment (pbougue) [inline source/kraken/worker.cpp:789]: indent

File: source/ptreferential/ptreferential_api.cpp [hunk 25]
```diff
@@ -41,124 +41,132 @@ namespace pt = boost::posix_time;
...
+        pb_creator.pb_fill(data.get_data<nt::Dataset>(rows), depth);
+        return;
+    default: return;
+    }
 }
 
 
-pbnavitia::Response query_pb(const type::Type_e requested_type,
+void query_pb(navitia::PbCreator& pb_creator,
+                             const type::Type_e requested_type,
```
Comment (pbougue) [inline source/ptreferential/ptreferential_api.cpp:140]: indent

File: source/time_tables/request_handle.cpp [hunk 26]
```diff
@@ -45,23 +45,23 @@ RequestHandle::RequestHandle(PbCreator& pb_creator,
     if (! calendar_id) {
         //we only have to check the production period if we do not have a calendar,
         // since if we have one we are only interested in the time, not the date
-        if(! pb_creator.data.meta->production_date.contains(datetime.date()) ) {
+        if(! pb_creator.data->meta->production_date.contains(datetime.date()) ) {
             pb_creator.fill_pb_error(pbnavitia::Error::date_out_of_bounds, "date is out of bound");
-        } else if( !pb_creator.data.meta->production_date.contains((datetime + boost::posix_time::seconds(duration)).date()) ) {
+        } else if( !pb_creator.data->meta->production_date.contains((datetime + boost::posix_time::seconds(duration)).date()) ) {
```
Comment (pbougue) [inline source/time_tables/request_handle.cpp:50]: line length

File: source/time_tables/request_handle.cpp [hunk 27]
```diff
@@ -45,23 +45,23 @@ RequestHandle::RequestHandle(PbCreator& pb_creator,
     if (! calendar_id) {
         //we only have to check the production period if we do not have a calendar,
         // since if we have one we are only interested in the time, not the date
-        if(! pb_creator.data.meta->production_date.contains(datetime.date()) ) {
+        if(! pb_creator.data->meta->production_date.contains(datetime.date()) ) {
```
Comment (pbougue) [inline source/time_tables/request_handle.cpp:48]: ) ) spacing

File: source/calendar/calendar_api.cpp (hunk 16 above)
Comment (TeXitoi) [reply source/calendar/calendar_api.cpp:39]: current_dt is in pb_convertor, and is used only in the function for pb_creator creation.

File: source/kraken/tests/realtime_test.cpp (hunk 19 above)
Comment (TeXitoi) [reply source/kraken/tests/realtime_test.cpp:1480]: I don't understand your remark.

File: source/kraken/worker.cpp (hunk 20 above)
Comment (TeXitoi) [reply source/kraken/worker.cpp:307]: That's already the name of the member.

File: source/kraken/worker.cpp (hunk 22 above)
Comment (TeXitoi) [reply source/kraken/worker.cpp:453]: it's in the pb_creator.

File: source/jormungandr/tests/ptref_tests.py (hunk 18 above)
Comment (TeXitoi) [reply source/jormungandr/tests/ptref_tests.py:95]:
```c++
std::set<const nt::Contributor*> contributors;
```
I'll open another PR to handle that.

File: source/kraken/tests/realtime_test.cpp (hunk 19 above)
Comment (pbougue) [reply source/kraken/tests/realtime_test.cpp:1480]: Weird to use both current and requested dt

**Pull request merged**